<a href="https://colab.research.google.com/github/tahae02/Week-1-Data-Analysis-Internship-Uptrail/blob/main/Week_1_Data_Analysis_for_Business_Insights.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#This is the notebook for the Week 1 project

In [ ]:
# from google.colab import drive
# drive.mount('/content/drive')

##Importing the necessary library to import and read the dataset

To avoid reuploading the dataset for every new runtime, I saved the dataset to my drive under the file path mentioned below, and 'Mounted the drive'.

In [ ]:
import pandas as pd

# file_path="/content/drive/MyDrive/Colab Notebooks/Uptrail Data Internship Projects /sales_data.xlsx"
file_path="/content/sales_data.xlsx"

df=pd.read_excel(file_path)
pd.read_excel(file_path) # this duplicate line is just to see what the dataset looks like at the start before pre-processing

Transaction_ID        Date Customer_ID     Product     Category  Quantity  \
0             1001  2024-01-05        C001      Laptop  Electronics         1   
1             1002  2024-01-10        C002  Smartphone  Electronics         2   
2             1003  2024-01-12        C003  Headphones  Electronics         1   
3             1004  2024-02-05        C004      Tablet  Electronics         1   
4             1005  2024-02-08        C005        Book        Books         3   
5             1006  2024-02-10        C001      Laptop  Electronics         1   
6             1007  2024-03-15        C006       Shoes     Clothing         2   
7             1008  2024-03-18        C007     T-Shirt     Clothing         1   
8             1009  2024-03-20        C008  Smartwatch  Electronics         1   
9             1010  2024-04-01        C009        Book        Books         2   
10            1011  2024-04-05        C002  Smartphone  Electronics         2   
11            1012  2024-04-10        C010      Tablet  Electronics         1   
12            1013  2024-05-01        C011       Shoes     Clothing         1   
13            1014  2024-05-05        C012  Headphones  Electronics         1   
14            1015  2024-05-08        C013      Laptop  Electronics         1   
15            1016  2024-05-10        C014     T-Shirt     Clothing         3   
16            1017  2024-06-01        C015  Smartwatch  Electronics         1   
17            1018  2024-06-05        C016        Book        Books         4   
18            1019  2024-06-08        C017  Smartphone  Electronics         1   
19            1020  2024-06-10        C018      Tablet  Electronics         1   

    Price  Total_Amount Payment_Method Region  
0     800           NaN    Credit Card  North  
1     600        1200.0           Cash  South  
2     100         100.0         PayPal   West  
3     500         500.0     Debit Card   East  
4      20          60.0    Credit Card  North  
5     800         800.0    Credit Card  North  
6      50         100.0           Cash  South  
7      25          25.0         PayPal   West  
8     200         200.0     Debit Card   East  
9      20          40.0    Credit Card  North  
10    600        1200.0           Cash  South  
11    500         500.0     Debit Card   East  
12     50          50.0           Cash  South  
13    100         100.0         PayPal   West  
14    800         800.0    Credit Card  North  
15     25          75.0           Cash  South  
16    200         200.0     Debit Card   East  
17     20          80.0    Credit Card  North  
18    600         600.0           Cash  South  
19    500         500.0     Debit Card   East

##Pre-Processing the Dataset

First I will look for duplicate values in the dataset. In this case, even though there are duplicates in some columns, there are no columns where all columns are same, so there are no true duplicates.


Next I will look at the dataset and find out how many null values there are that I need to handle

In [ ]:
df.isnull().sum() # this code goes through every column and gives me the count of null values for each column

Transaction_ID    0
Date              0
Customer_ID       0
Product           0
Category          0
Quantity          0
Price             0
Total_Amount      1
Payment_Method    0
Region            0
dtype: int64

<!-- Next I will try removing that null value, since it is a number i can fill it in with a statisical value (I will try this later) -->

You can see there is 1 NULL value in the 'Total_Amount' column.

One way I can handle this missing value is by dropping the entire row.

In [ ]:
# df.dropna()

In this case the missing value is a numerical value, so removing the row completely is not ideal, the preferrable solution would be to fill in the cell with a column-related statistical value, for example the mean.

In [ ]:
mean_total_amount=df['Total_Amount'].mean()
print(f"\nMean of 'Total_Amount': {mean_total_amount}")
print()

df['Total_Amount'].fillna(mean_total_amount, inplace=True)

print(df) # new dataset with no NULL values


Mean of 'Total_Amount': 375.2631578947368

    Transaction_ID        Date Customer_ID     Product     Category  Quantity  \
0             1001  2024-01-05        C001      Laptop  Electronics         1   
1             1002  2024-01-10        C002  Smartphone  Electronics         2   
2             1003  2024-01-12        C003  Headphones  Electronics         1   
3             1004  2024-02-05        C004      Tablet  Electronics         1   
4             1005  2024-02-08        C005        Book        Books         3   
5             1006  2024-02-10        C001      Laptop  Electronics         1   
6             1007  2024-03-15        C006       Shoes     Clothing         2   
7             1008  2024-03-18        C007     T-Shirt     Clothing         1   
8             1009  2024-03-20        C008  Smartwatch  Electronics         1   
9             1010  2024-04-01        C009        Book        Books         2   
10            1011  2024-04-05        C002  Smartphone  Electroni

<ipython-input-8-f3b1c20fab5c>:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Total_Amount'].fillna(mean_total_amount, inplace=True)


Now you can see that the first value in 'Total_Amount' is no longer 'NaN', it is 375.263158

## Next step is to standardise the columns
Ensuring:

 all dates are in the format 'YYYY/mm/DD' for trend analysis

numerical values do not have commas or any other symbols



I will also drop some columns that are not needed for analysis or visualisation

this helps keep the dataset cleaner

In [ ]:
df.drop(columns=['Transaction_ID', 'Customer_ID'], inplace=True)

Once all the cleaning is complete, I need to export the new clean dataset

In [ ]:
df.to_csv("cleaned_sales_data.csv", index=False)

#2. Exploratory Data Analysis

opening the cleaned dataset and outputting the structure

In [ ]:
new_file_path="/content/cleaned_sales_data.csv"
df_cleaned=pd.read_csv(new_file_path)

df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            20 non-null     object 
 1   Product         20 non-null     object 
 2   Category        20 non-null     object 
 3   Quantity        20 non-null     int64  
 4   Price           20 non-null     int64  
 5   Total_Amount    20 non-null     float64
 6   Payment_Method  20 non-null     object 
 7   Region          20 non-null     object 
dtypes: float64(1), int64(2), object(5)
memory usage: 1.4+ KB


Looking at the first 5 rows of the new cleaned dataset

In [ ]:
df_cleaned.head()

Date     Product     Category  Quantity  Price  Total_Amount  \
0  2024-01-05      Laptop  Electronics         1    800    375.263158   
1  2024-01-10  Smartphone  Electronics         2    600   1200.000000   
2  2024-01-12  Headphones  Electronics         1    100    100.000000   
3  2024-02-05      Tablet  Electronics         1    500    500.000000   
4  2024-02-08        Book        Books         3     20     60.000000   

  Payment_Method Region  
0    Credit Card  North  
1           Cash  South  
2         PayPal   West  
3     Debit Card   East  
4    Credit Card  North

understanding the new dataset

displaying statistical data for the numerical columns

In [ ]:
df_cleaned[['Quantity', 'Price', 'Total_Amount']].describe()

Quantity       Price  Total_Amount
count  20.000000   20.000000     20.000000
mean    1.550000  325.500000    375.263158
std     0.887041  302.484884    378.690520
min     1.000000   20.000000     25.000000
25%     1.000000   43.750000     78.750000
50%     1.000000  200.000000    200.000000
75%     2.000000  600.000000    525.000000
max     4.000000  800.000000   1200.000000

trying to find the link between time of year and total amount

the most effective way of doing this is grouping the dates by month.

In [ ]:
df_cleaned['Date'] = pd.to_datetime(df_cleaned['Date']) # converting to datetime

# creating a new column for the month
df_cleaned['Month'] = df_cleaned['Date'].dt.to_period('M')  # gives '2024-01', '2024-02', etc.
df_cleaned.head()


Date     Product     Category  Quantity  Price  Total_Amount  \
0 2024-01-05      Laptop  Electronics         1    800    375.263158   
1 2024-01-10  Smartphone  Electronics         2    600   1200.000000   
2 2024-01-12  Headphones  Electronics         1    100    100.000000   
3 2024-02-05      Tablet  Electronics         1    500    500.000000   
4 2024-02-08        Book        Books         3     20     60.000000   

  Payment_Method Region    Month  
0    Credit Card  North  2024-01  
1           Cash  South  2024-01  
2         PayPal   West  2024-01  
3     Debit Card   East  2024-02  
4    Credit Card  North  2024-02